### PR TITLE
Fix dead lock in`ComputeServerRunner.*`

### LINKED ISSUE #7410: Tsan: `ComputeServerRunner.*` dead lock
https://ci.pingcap.net/blue/rest/organizations/jenkins/pipelines/tiflash-sanitizer-daily/runs/1073/nodes/85/steps/89/log/?start=0

```
[2023-05-03T18:07:39.142Z] WARNING: ThreadSanitizer: lock-order-inversion (potential deadlock) (pid=6218)
[2023-05-03T18:07:39.142Z]   Cycle in lock order graph: M896351926563439496 (0x000000000000) => M764058687509430424 (0x000000000000) => M896351926563439496
[2023-05-03T18:07:39.142Z] 
[2023-05-03T18:07:39.142Z]   Mutex M764058687509430424 acquired here while holding mutex M896351926563439496 in thread T2296:
[2023-05-03T18:07:39.142Z]     #0 pthread_mutex_lock /llvm-project/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_common_interceptors.inc:4249:3 (gtests_dbms+0x27cd718)
[2023-05-03T18:07:39.142Z]     #1 std::__1::mutex::lock() <null> (libc++.so.1+0x66f45)
[2023-05-03T18:07:39.142Z]     #2 DB::ExchangeReceiverBase<DB::GRPCReceiverContext>::createAsyncRequestHandler(DB::ExchangeRecvRequest&&)::'lambda0'(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&)::operator()(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&) const /home/jenkins/agent/workspace/tiflash-sanitizer-daily/tiflash/dbms/src/Flash/Mpp/ExchangeReceiver.cpp:566:27 (gtests_dbms+0x2b17eb6)
[2023-05-03T18:07:39.142Z]     #3 decltype(static_cast<DB::ExchangeReceiverBase<DB::GRPCReceiverContext>::createAsyncRequestHandler(DB::ExchangeRecvRequest&&)::'lambda0'(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&)&>(fp)(static_cast<bool>(fp0), static_cast<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>(fp0), static_cast<std::__1::shared_ptr<DB::Logger> const&>(fp0))) std::__1::__invoke<DB::ExchangeReceiverBase<DB::GRPCReceiverContext>::createAsyncRequestHandler(DB::ExchangeRecvRequest&&)::'lambda0'(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&)&, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&>(DB::ExchangeReceiverBase<DB::GRPCReceiverContext>::createAsyncRequestHandler(DB::ExchangeRecvRequest&&)::'lambda0'(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&)&, bool&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&) /usr/local/bin/../include/c++/v1/type_traits:3918:1 (gtests_dbms+0x2b17eb6)
[2023-05-03T18:07:39.142Z]     #4 void std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::ExchangeReceiverBase<DB::GRPCReceiverContext>::createAsyncRequestHandler(DB::ExchangeRecvRequest&&)::'lambda0'(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&)&, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&>(DB::ExchangeReceiverBase<DB::GRPCReceiverContext>::createAsyncRequestHandler(DB::ExchangeRecvRequest&&)::'lambda0'(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&)&, bool&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&) /usr/local/bin/../include/c++/v1/__functional/invoke.h:61:9 (gtests_dbms+0x2b17eb6)
[2023-05-03T18:07:39.142Z]     #5 std::__1::__function::__alloc_func<DB::ExchangeReceiverBase<DB::GRPCReceiverContext>::createAsyncRequestHandler(DB::ExchangeRecvRequest&&)::'lambda0'(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&), std::__1::allocator<DB::ExchangeReceiverBase<DB::GRPCReceiverContext>::createAsyncRequestHandler(DB::ExchangeRecvRequest&&)::'lambda0'(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&)>, void (bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&)>::operator()(bool&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&) /usr/local/bin/../include/c++/v1/__functional/function.h:171:16 (gtests_dbms+0x2b17eb6)
[2023-05-03T18:07:39.142Z]     #6 std::__1::__function::__func<DB::ExchangeReceiverBase<DB::GRPCReceiverContext>::createAsyncRequestHandler(DB::ExchangeRecvRequest&&)::'lambda0'(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&), std::__1::allocator<DB::ExchangeReceiverBase<DB::GRPCReceiverContext>::createAsyncRequestHandler(DB::ExchangeRecvRequest&&)::'lambda0'(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&)>, void (bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&)>::operator()(bool&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&) /usr/local/bin/../include/c++/v1/__functional/function.h:345:12 (gtests_dbms+0x2b17eb6)
[2023-05-03T18:07:39.142Z]     #7 std::__1::__function::__value_func<void (bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&)>::operator()(bool&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&) const /usr/local/bin/../include/c++/v1/__functional/function.h:498:16 (gtests_dbms+0x2b18e84)
[2023-05-03T18:07:39.142Z]     #8 std::__1::function<void (bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&)>::operator()(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Logger> const&) const /usr/local/bin/../include/c++/v1/__functional/function.h:1175:12 (gtests_dbms+0x2b18e84)
[2023-05-03T18:07:39.142Z]     #9 DB::AsyncRequestHandler<DB::GRPCReceiverContext, false>::closeConnection(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&) /home/jenkins/agent/workspace/tiflash-sanitizer-daily/tiflash/dbms/src/Flash/Mpp/AsyncRequestHandler.h:276:13 (gtests_dbms+0x2b18e84)
[2023-05-03T18:07:39.142Z]     #10 DB::AsyncRequestHandler<DB::GRPCReceiverContext, false>::processWaitFinish() /home/jenkins/agent/workspace/tiflash-sanitizer-daily/tiflash/dbms/src/Flash/Mpp/AsyncRequestHandler.h:174:13 (gtests_dbms+0x2b18be9)
[2023-05-03T18:07:39.142Z]     #11 DB::AsyncRequestHandler<DB::GRPCReceiverContext, false>::execute(bool&) /home/jenkins/agent/workspace/tiflash-sanitizer-daily/tiflash/dbms/src/Flash/Mpp/AsyncRequestHandler.h:123:17 (gtests_dbms+0x2b1806e)
[2023-05-03T18:07:39.142Z]     #12 DB::GRPCCompletionQueuePool::thread(unsigned long) /home/jenkins/agent/workspace/tiflash-sanitizer-daily/tiflash/dbms/src/Flash/Mpp/GRPCCompletionQueuePool.cpp:63:48 (gtests_dbms+0xdb75596)
[2023-05-03T18:07:39.142Z] [92m[2023/05/04 02:07:30.864 +08:00] [0m[1m[90m[TRACE] [0m[35m[TiRemoteBlockInputStream.h:106] [0m["recv 250 rows from remote for tunnel2+-1, total recv row num: 4250"] [0m[90m[source=TiRemote(ExchangeReceiver)] [thread_id=1][0m
```

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #7410

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Mpp/AsyncRequestHandler.h [hunk 1]
```diff
@@ -137,6 +137,7 @@ class AsyncRequestHandler : public AsyncRequestHandlerBase
     {
         std::unique_lock<std::mutex> ul(is_close_conn_called_mu);
         condition_cv.wait(ul, [this]() { return is_close_conn_called; });
+        ul.unlock();
```
Comment (windtalker) [inline dbms/src/Flash/Mpp/AsyncRequestHandler.h:140]: the destructor of `ul` will unlock it, so why call unlock explicitly here?

File: dbms/src/Flash/Mpp/ExchangeReceiver.cpp [hunk 2]
```diff
@@ -409,6 +409,8 @@ void ExchangeReceiverBase<RPCContext>::waitAllConnectionDone()
     // In order to ensure the destructions of local tunnels are
     // after the ExchangeReceiver, we need to wait at here.
     waitLocalConnectionDone(lock);
+
+    lock.unlock();
```
Comment (windtalker) [inline dbms/src/Flash/Mpp/ExchangeReceiver.cpp:413]: Can you add some comments to explain why `waitLocalConnectionDone` need lock and `waitAsyncConnectionDone` don't need lock?

File: dbms/src/Flash/Mpp/AsyncRequestHandler.h (hunk 1 above)
Comment (xzhangxian1008) [reply dbms/src/Flash/Mpp/AsyncRequestHandler.h:140]: > the destructor of `ul` will unlock it, so why call unlock explicitly here?

oh, I ignore the destructor, and i will delete it.

File: dbms/src/Flash/Mpp/ExchangeReceiver.cpp (hunk 2 above)
Comment (xzhangxian1008) [reply dbms/src/Flash/Mpp/ExchangeReceiver.cpp:419]: > Can you add some comments to explain why `waitLocalConnectionDone` need lock and `waitAsyncConnectionDone` don't need lock?

okk

Comment (xzhangxian1008) [reply dbms/src/Flash/Mpp/ExchangeReceiver.cpp:419]: > why not
> 
> ```
> {
>     std::unique_lock lock(mu);
>     ...
> }
> waitAsyncConnectionDone();
> ```
> 
> ?

okk

**Review by SeaRise — APPROVED** [review]

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**